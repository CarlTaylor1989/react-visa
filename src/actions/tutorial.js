import * as types from './tutorialTypes';

export const setTutorialSlideIndex = slideIndex => ({
  type: types.SET_TUTORIAL_SLIDE_INDEX,
  slideIndex
});

export const setTutorialLaunched = launched => ({
  type: types.SET_TUTORIAL_LAUNCHED,
  launched
});
