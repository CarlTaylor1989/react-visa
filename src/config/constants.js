// Loader
export const REMOVE_LOADER_AFTER = 2000;

// Diagnostic
export const DIAGNOSTIC_NOT_STARTED = 0;
export const DIAGNOSTIC_STARTED = 1;
export const DIAGNOSTIC_COMPLETED = 2;

// Client statuses
export const CLIENT_NOT_STARTED = 0;
export const CLIENT_IN_PROGRESS = 1;
export const CLIENT_COMPLETED = 2;
export const CLIENT_LOCKED = 3;
export const CLIENT_GIVENUP = 4;

// Family statuses
export const FAMILY_INCOMPLETE_STATUS = 0;
export const FAMILY_COMPLETION_STATUS = 1;

// Challenge statuses
export const CHALLENGE_UNAVAILABLE_STATUS = -1;
export const CHALLENGE_AVAILABLE_STATUS = 0;
export const CHALLENGE_IN_PROGRESS_STATUS = 1;
export const CHALLENGE_FAILED_STATUS = 2;
export const CHALLENGE_COMPLETION_STATUS = 3;
export const CHALLENGE_PERFECT_STATUS = 4;

// Challenge CSS status classes
export const CHALLENGE_NOT_STARTED_CLASS = 'notStarted';
export const CHALLENGE_AVAILABLE_CLASS = 'available';
export const CHALLENGE_COMPLETED_CLASS = 'completed';
export const CHALLENGE_PERFECT_CLASS = 'perfect';

// Product CSS status classes
export const PRODUCT_INCOMPLETE_CLASS = 'incomplete';
export const PRODUCT_COMPLETED_CLASS = 'completed';

// SCO completion statuses and perfect score
export const SCO_PERFECT_SCORE = 100;
export const SCO_COMPLETION_PASSED = 'passed';
export const SCO_COMPLETION_FAILED = 'failed';
export const SCO_COMPLETION_UNKNOWN = 'unknown';

// Streak bonus
export const ACTIVATE_STREAK_AFTER = 14; // hours
export const MAXIMUM_STREAK = 4;

// Prompt types
export const ACHIEVEMENT_PROMPT = 'achievement';
export const RANK_PROMPT = 'rank';
export const PRODUCT_PROMPT = 'product';
export const BONUS_STREAK_PROMPT = 'bonusStreak';
export const EXIT_SUMMARY_PROMPT_AFTER = 5000;

// Consecutive days
export const CONSECUTIVE_LOWER = 14;
export const CONSECUTIVE_UPPER = 33;

// Tooltip delay
export const DISPLAY_TOOLTIPS_AFTER = 500; // miliseconds

// xAPI
export const XAPI_CHALLENGE_MIN_SCORE = 0;
export const XAPI_CHALLENGE_MAX_SCORE = 2;
export const XAPI_FAIL_CHALLENGE = 'fail';
export const XAPI_FAIL_CHALLENGE_SCORE = 0;
export const XAPI_COMPLETED_CHALLENGE = 'pass';
export const XAPI_COMPLETED_CHALLENGE_SCORE = 1;
export const XAPI_PERFECT_CHALLENGE = 'perfect';
export const XAPI_PERFECT_CHALLENGE_SCORE = 2;
export const XAPI_FEEDBACK_MIN_RATING = 1;
export const XAPI_FEEDBACK_MAX_RATING = 5;

// Cornerstone LMS URL
export const LMS_URL = 'https://visalms.csod.com/LMS/scorm/';

// Feedback
export const FEEDBACK_TRIGGER = 3;
