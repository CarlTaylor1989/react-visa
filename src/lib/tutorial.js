import {
  networkMapPath,
  networkProductPath,
  clientsSelectorPath,
  proposalPath
} from '../config/navigation';
import tutorialIds from '../config/tutorials';

export const lib = {};
/**
 * Based on the location, it finds and returns the tutorial ID
 * @param {string} location The current game location
 * @return {string}
 */
lib.findTutorialId = (location) => {
  let tutorialId = '';
  switch (location) {
    case networkMapPath:
    case networkProductPath:
      tutorialId = 't4';
      break;
    case clientsSelectorPath:
    case proposalPath:
      tutorialId = 't3';
      break;
    default:
      tutorialId = 't1';
      break;
  }
  return tutorialId;
};

/**
 * Get the tutorial slide index
 * @param {string} id Tutorial ID
 * @return {number}
 */
lib.getTutorialSlideIndex = id => (tutorialIds[id] ? tutorialIds[id].slide : 0);

/**
 * Find and gets the tutorial slide index based on a given location
 * @param {string} location The current game location
 */
export const getLocationTutorialSlideIndex = (location) => {
  const tutorialId = lib.findTutorialId(location);
  return lib.getTutorialSlideIndex(tutorialId);
};
