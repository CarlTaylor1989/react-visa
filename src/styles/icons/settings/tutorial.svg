<svg xmlns="http://www.w3.org/2000/svg" width="48.096" height="35.345" viewBox="0 0 48.096 35.345">
  <defs>
    <style>
      .cls-1 {
        fill: #fff;
      }
    </style>
  </defs>
  <g id="Group_1020" data-name="Group 1020" transform="translate(-374.082 -350.371)">
    <rect id="Rectangle_564" data-name="Rectangle 564" class="cls-1" width="48.096" height="28.653" rx="5.877" transform="translate(374.082 350.371)"/>
    <path id="Path_1734" data-name="Path 1734" class="cls-1" d="M383.892,373.134l-7.231,8.838s-2.008,2.41.4,3.616c0,0,1.205.8,5.222-1.607l12.453-5.624Z"/>
  </g>
</svg>
