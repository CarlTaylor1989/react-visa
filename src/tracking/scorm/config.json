{
	"version": "2004 3rd Edition",
	"debugMode": false
}
