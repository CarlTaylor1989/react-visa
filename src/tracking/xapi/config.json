{
  "version": "1.0.0",
  "projectId": "mario19-037",
  "gameName": "visa-inside-track",
  "baseURL": "https://lms.visa.com/e-learning/",
  "extensionURL": "https://lms.visa.com/xapi/cornerstone/extensions/",
  "title": "Mario technical prototype",
  "homePage": "https://lms.visa.com",
  "gomoCourseURL": "http://www.gomolearning.com/xapi/activities/course/", 
  "debugMode": false
}
