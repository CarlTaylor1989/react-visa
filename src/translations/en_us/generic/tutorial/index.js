import * as generic from './generic.json';
import * as tutorials from './tutorials.json';

export default {
  ...generic,
  ...tutorials
};
