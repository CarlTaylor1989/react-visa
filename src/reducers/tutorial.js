import { createReducer } from 'redux-create-reducer';
import * as types from '../actions/tutorialTypes';

export const initialState = {
  slideIndex: 0,
  launched: false
};

export const tutorialData = createReducer(Object.assign(initialState), {
  [types.SET_TUTORIAL_SLIDE_INDEX](state, action) {
    return {
      ...state,
      slideIndex: action.slideIndex
    };
  },
  [types.SET_TUTORIAL_LAUNCHED](state, action) {
    return {
      ...state,
      launched: action.launched
    };
  }
});

export default { tutorialData };
