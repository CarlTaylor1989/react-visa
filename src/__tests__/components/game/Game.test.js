/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import {
  Game,
  mapStateToProps
} from '../../../components/game/Game';
import { getLocationTutorialSlideIndex } from '../../../lib/tutorial';
import { FEEDBACK_TRIGGER } from '../../../config/constants';

jest.mock('../../../lib/tutorial');

let mockProps;
beforeEach(() => {
  jest.resetAllMocks();
  mockProps = {
    commit: jest.fn(),
    diagnosticCompleted: true,
    disableFeedback: jest.fn(),
    feedbackTriggered: false,
    feedbackDisabled: false,
    feedbackDays: 0,
    lessonLocation: '',
    location: {
      pathname: '/irrelevant'
    },
    prepareSuspendData: jest.fn(),
    push: jest.fn(),
    setLessonLocation: jest.fn(),
    setPlayAudio: jest.fn(),
    setScreenReferrer: jest.fn(),
    setTutorialSlideIndex: jest.fn(),
    setGenericData: jest.fn(),

  };
});

test('renders the Game component', () => {
  const wrapper = shallow(<Game {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders the Game component with the feedback popup', () => {
  const wrapper = shallow(<Game
    {...mockProps}
    feedbackTriggered
    feedbackDays={FEEDBACK_TRIGGER}
  />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('method componentDidMount()', () => {
  // setup
  const wrapper = shallow(<Game {...mockProps} />, {
    disableLifecycleMethods: true
  });
  wrapper.instance().onDisplayFeedbackPopup = jest.fn();
  wrapper.instance().state.displayFeedbackPopup = true;

  // function under test
  wrapper.instance().componentDidMount();

  // expectation
  expect(mockProps.disableFeedback).toHaveBeenCalled();
  expect(wrapper.instance().onDisplayFeedbackPopup).toHaveBeenCalled();
});

test('method isNetworkBtnActive()', () => {
  // setup
  const wrapper = shallow(<Game {...mockProps} />);
  wrapper.instance().checkForChildPaths = jest.fn().mockReturnValue(true);

  // function under test
  const output = wrapper.instance().isNetworkBtnActive();

  // expectation
  expect(output).toBeTruthy();
});

test('method isClientBtnActive()', () => {
  // setup
  const wrapper = shallow(<Game {...mockProps} />);
  wrapper.instance().checkForChildPaths = jest.fn().mockReturnValue(true);

  // function under test
  const output = wrapper.instance().isClientBtnActive();

  // expectation
  expect(output).toBeTruthy();
});

test('method checkForChildPaths()', () => {
  // setup
  const pathname = 'irrelevant';
  mockProps.location = { pathname };
  const wrapper = shallow(<Game {...mockProps} />);

  // function under test
  const output = wrapper.instance().checkForChildPaths(pathname);

  // expectation
  expect(output).toBeTruthy();
});

test('method onCloseHelpPopup()', () => {
  const wrapper = shallow(<Game {...mockProps} />);
  wrapper.instance().onCloseHelpPopup();
  expect(wrapper.instance().state.displayHelpPopup).toBeFalsy();
});

test('method onCloseFeedbackPopup()', () => {
  const wrapper = shallow(<Game {...mockProps} />);
  wrapper.instance().onCloseFeedbackPopup();
  expect(wrapper.instance().state.displayFeedbackPopup).toBeFalsy();
});

test('method onDisplayHelpPopup()', () => {
  // setup
  getLocationTutorialSlideIndex.mockImplementation(() => 3);
  const wrapper = shallow(<Game {...mockProps} />);

  // function under test
  wrapper.instance().onDisplayHelpPopup();

  // expectations
  expect(mockProps.setTutorialSlideIndex).toHaveBeenCalledWith(expect.any(Number));
  expect(wrapper.instance().state.displayHelpPopup).toBeTruthy();
});

test('method onDisplayFeedbackPopup()', () => {
  const wrapper = shallow(<Game {...mockProps} />);
  wrapper.instance().onDisplayFeedbackPopup();
  expect(wrapper.instance().state.displayFeedbackPopup).toBeTruthy();
});

test('should mapStateToProps', () => {
  const state = {
    genericData: {
      feedback: {
        triggered: false
      }
    },
  };
  const props = mapStateToProps(state);
  expect(props).toEqual({
    feedbackTriggered: state.genericData.feedback.triggered,
    feedbackDays: state.genericData.feedback.days
  });
});
