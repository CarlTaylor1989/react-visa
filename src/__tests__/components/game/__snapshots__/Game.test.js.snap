// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Game component 1`] = `
<div
  className="game"
>
  <div
    className="logo"
  />
  <div
    className="decor stars"
  />
  <div
    className="decor circlesTop"
  />
  <div
    className="decor circlesBottom"
  />
  <div
    className="vuLogo"
  />
  <div
    className="navButtons"
  >
    <NavLink
      activeClassName="selected"
      className="clientPropsBtn"
      onMouseEnter={[Function]}
      onTouchEnd={[Function]}
      to="/game/clientproposals"
    >
      <span
        className="icon"
      />
      <Translate
        tag="span"
        value="clientprops.clientPropsBtn"
      />
    </NavLink>
    <NavLink
      activeClassName="selected"
      className="productNetworkBtn"
      isActive={[Function]}
      onMouseEnter={[Function]}
      onTouchEnd={[Function]}
      to="/game/network/map"
    >
      <span
        className="icon"
      />
      <Translate
        tag="span"
        value="network.productNetworkBtn"
      />
    </NavLink>
  </div>
  <button
    className="helpBtn"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchEnd={[Function]}
    type="button"
  >
    <Translate
      className="text"
      tag="span"
      value="generic.helpBtn"
    />
    <span
      className="icon"
    />
  </button>
  <Switch>
    <Route
      component={[Function]}
      path="/game/network"
    />
    <Route
      component={[Function]}
      path="/game/clientproposals"
    />
    <Redirect
      to="/game/clientproposals"
    />
  </Switch>
  <Connect(MiniDashboard) />
  <Connect(TutorialPopup)
    closePopup={[Function]}
    showPopup={false}
  />
  <FeedbackPopup
    closePopup={[Function]}
    showPopup={false}
  />
  <Connect(FooterBar)
    changeRegion={[Function]}
  />
  <Connect(Prompts) />
</div>
`;

exports[`renders the Game component with the feedback popup 1`] = `
<div
  className="game"
>
  <div
    className="logo"
  />
  <div
    className="decor stars"
  />
  <div
    className="decor circlesTop"
  />
  <div
    className="decor circlesBottom"
  />
  <div
    className="vuLogo"
  />
  <div
    className="navButtons"
  >
    <NavLink
      activeClassName="selected"
      className="clientPropsBtn"
      onMouseEnter={[Function]}
      onTouchEnd={[Function]}
      to="/game/clientproposals"
    >
      <span
        className="icon"
      />
      <Translate
        tag="span"
        value="clientprops.clientPropsBtn"
      />
    </NavLink>
    <NavLink
      activeClassName="selected"
      className="productNetworkBtn"
      isActive={[Function]}
      onMouseEnter={[Function]}
      onTouchEnd={[Function]}
      to="/game/network/map"
    >
      <span
        className="icon"
      />
      <Translate
        tag="span"
        value="network.productNetworkBtn"
      />
    </NavLink>
  </div>
  <button
    className="helpBtn"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchEnd={[Function]}
    type="button"
  >
    <Translate
      className="text"
      tag="span"
      value="generic.helpBtn"
    />
    <span
      className="icon"
    />
  </button>
  <Switch>
    <Route
      component={[Function]}
      path="/game/network"
    />
    <Route
      component={[Function]}
      path="/game/clientproposals"
    />
    <Redirect
      to="/game/clientproposals"
    />
  </Switch>
  <Connect(MiniDashboard) />
  <Connect(TutorialPopup)
    closePopup={[Function]}
    showPopup={false}
  />
  <FeedbackPopup
    closePopup={[Function]}
    showPopup={true}
  />
  <Connect(FooterBar)
    changeRegion={[Function]}
  />
  <Connect(Prompts) />
</div>
`;
